[pytorch][codegen] migrate gen_variable_factories.py to the new data model

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47818 [pytorch][codegen] migrate gen_variable_factories.py to the new data model**
* #47747 [pytorch][codegen] add fully migrated scripts to mypy strict config
* #47746 [pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly
* #47745 [pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model

This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

ljk53 committed:
[pytorch][codegen] migrate gen_variable_factories.py to the new data model

This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

[ghstack-poisoned]

bhosmer commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -3,77 +3,79 @@
+        else:
+            exprs.append(arg.name)
+
     requires_grad = "options.requires_grad()" if has_tensor_options else "false"
```
...also, does the use of `options.requires_grad()` here contradict the comment above about ignoring it? Both are in old code, I know... just wondering if there's an easy clarification to be had. But NBD if not

ljk53 replied:
> ...also, does the use of `options.requires_grad()` here contradict the comment above about ignoring it? Both are in old code, I know... just wondering if there's an easy clarification to be had. But NBD if not

The comment above is talking about the `options` argument being passed into aten, which is ignored. The use of `options.requires_grad()` here is to set the requires grad bit on the result tensor. See the example:
```
inline at::Tensor empty(at::IntArrayRef size, const at::TensorOptions & options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  at::Tensor tensor = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::empty(size, at::TensorOptions(options).requires_grad(c10::nullopt), memory_format);
  })();
  at::Tensor result =
    autograd::make_variable(std::move(tensor), /*requires_grad=*/options.requires_grad());
  return result;
}
```

bhosmer replied:
thanks! your explanation actually improves on the comment :) but not worth further pre-landing churn on your part

ljk53 replied:
> thanks! your explanation actually improves on the comment :) but not worth further pre-landing churn on your part

Added some comment - having both handled in `isinstance(arg.argument, TensorOptionsArguments)` branch it should be more clear what we are trying to do.

bhosmer reviewed (approved):
Looks good! Handful of minor thoughts but nothing huge - only one non-style thing (the hardcoded "options" arg name).

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_factories.py to the new data model"


This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

[ghstack-poisoned]

ljk53 commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -3,77 +3,79 @@
+    native_functions = parse_native_yaml(native_yaml_path)
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
+    fm.write_with_template('variable_factories.h', 'variable_factories.h', lambda: {
+        'generated_comment': f'@generated from {fm.template_dir}/variable_factories.h',
```
> just curious why the override is necessary here... what would `write_with_template()` be plugging in for `generated_comment` by default?

The default comment is saying the file is generated by `aten/src/ATen/gen.py`

bhosmer replied:
ah got it... lol seems like that defaulting could be improved but obvs not for this PR

ljk53 commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -3,77 +3,79 @@
-          CodeTemplate.from_file(template_path + "/variable_factories.h"),
-          {"function_definitions": function_definitions})
+    if Variant.function not in f.variants or \
+            not has_tensor_options and not name.endswith("_like"):
```
done

ljk53 commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -3,77 +3,79 @@
+        return f'c10::optional<{type}>' if opt else type
 
     opt_match = OPTIONAL_TYPE_PATTERN.match(argument_type)
     if opt_match:
```
done

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_factories.py to the new data model"


This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_factories.py to the new data model"


This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_factories.py to the new data model"


This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_factories.py to the new data model"


This is another relatively small codegen.

Ideally we should CppSignature.decl() to generate the c++ function declaration.
We didn't because it needs to add 'at::' to the types defined in ATen namespace.

E.g.:
- standard declaration:
```
Tensor eye(int64_t n, int64_t m, const TensorOptions & options={})
```

- expected:
```
at::Tensor eye(int64_t n, int64_t m, const at::TensorOptions & options = {})
```

Kept the hacky fully_qualified_type() method to keep compatibility with old codegen.

We could clean up by:
- Using these types in torch namespace - but this is a user facing header file,
  not sure if it will cause problem;
- Update cpp.argument_type() method to take optional namespace argument;

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24909478](https://our.internmc.facebook.com/intern/diff/D24909478)

[ghstack-poisoned]
